[PyTorch] Migrate add operators to borrow in TensorIteratorBase

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55691 [PyTorch] Migrate add operators to borrow in TensorIteratorBase**
* #55690 [PyTorch] Support borrowing in/out Tensors in TensorIterator
* #55940 [PyTorch] Always update op.current_dtype in TensorIteratorBase::set_output
* #55420 [PyTorch] Remove non-const TensorIterator::tensor() method
* #55258 [PyTorch] Format generated structured kernels code better
* #55351 [PyTorch] Fix const correctness for resize native functions

Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

swolchok committed:
[PyTorch] Migrate add operators to borrow in TensorIteratorBase

Avoiding reference counting for these operations is very very
roughly a 5-6% win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -16,7 +16,7 @@ namespace meta {
   const Tensor& self, const Tensor& other, const Scalar& alpha
 ) {
-  build_binary_op(maybe_get_output(), self, other);
+  build_borrowing_binary_op(maybe_get_output(), self, other);
```
On the migration plan- some options sound like:

(1) Don't bother including both the borrowing and non-borrowing API's. Immediately opt all TensorIterator kernels to use borrowing

(2) Hitch this to structured kernels- all TI ops that are ported to structured will automatically use the borrowing interface

(3) Port kernels one-by-one to use borrowing (like in this PR).


Maybe (2) is a reasonable middle ground? (3) seems like too much work, but I could understand not doing (1), if we're worried about the API not getting used properly since we don't have a borrow checker

swolchok replied:
I'm not sure how to accomplish (2) concretely -- the code block we're commenting on is a structured kernel, and it required a manual one-by-one code change. How would you fix it?

bdhirsh replied:
One way is that unstructured and structured TI kernels have slightly different code paths. Structured kernels call into `TensorIteratorBase::build_binary_op` (or similar, for unary, etc), while unstructured kernels call `TensorIterator::binary_op` ([example](https://github.com/pytorch/pytorch/blob/f3367f917e916a9f57461089889f4f4ddbb16d20/aten/src/ATen/native/BinaryOps.cpp#L161)). Right now `binary_op` is written in terms if `build_binary_op`, but we could update it so that it doesn't use borrowing and `build_binary_op` does.

I'm not 100% sure on the benefit of only applying this to structured kernels though, aside from it being less of a "big bang" change. I doubt everyone who ports an op to be structured will know to think carefully about the borrowing logic and whether or not it would break for their op.

swolchok replied:
> I doubt everyone who ports an op to be structured will know to think carefully about the borrowing logic and whether or not it would break for their op.

Sounds like it's better to be explicit about what's going on, then.

ezyang commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -771,16 +771,28 @@ void TensorIteratorBase::build_binary_float_op(const Tensor& out, const Tensor&
 }
 
+// This cannot be a function because TensorIteratorConfig is not
+// copyable or movable, so it can't be returned from the function.
```
😳

swolchok committed:
rebase on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
fix newly-added resize_as_sparse_ on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
rebase w/conflicts on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
fix build at bottom of stack on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Migrate add operators to borrow in TensorIteratorBase"


Avoiding reference counting for these operations is
roughly a 5% CPU time win vs not supporting borrowing at all.

Differential Revision: [D27607295](https://our.internmc.facebook.com/intern/diff/D27607295/)

[ghstack-poisoned]
